fix(adapter-cloudflare-workers): throws status 404 if KVError instanceof

vicentematus opened the pull request:
# Describe the problem

Fixes #9288. The bug occurs when deploying to Cloudflare Workers, where requests for non-existing assets result in a 500 internal error instead of a 404 error. 

# The Solution

Add a `try-catch` clause to handle the exception if the `error` is an instance of [KVError from Cloudflare Workers](https://github.com/cloudflare/kv-asset-handler#getassetfromkv) it will change the `error.status` to `404 `instead of the default `500` as described in the PR.

---

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [X] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [X] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [X] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [X] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

benmccann commented on .changeset/large-guests-fix.md:
```diff
@@ -0,0 +1,5 @@
+'@sveltejs/adapter-cloudflare-workers': minor
+---
+
+fix: throws KVError when non-existing asset
```
this will show up in the changelog, so should be worded in a way that describes the new behavior rather than just the old behavior
```suggestion
fix: return 404 for non-existent asset rather than throwing KVError
```

Rich-Harris commented:
Thanks for the PR. It's not totally clear to me what effect setting `error.status = 404` is supposed to have — judging by https://github.com/cloudflare/kv-asset-handler#getassetfromkv we should be doing this instead:

```diff
async function get_asset_from_kv(req, env, context, map = mapRequestToAsset) {
	try {
		return await getAssetFromKV(
			{
				request: req,
				waitUntil(promise) {
					return context.waitUntil(promise);
				}
			},
			{
				ASSET_NAMESPACE: env.__STATIC_CONTENT,
				ASSET_MANIFEST: static_asset_manifest,
				mapRequestToAsset: map
			}
		);
	} catch (err) {
-		if (err instanceof KVError) err.status = 404;
+		if (e instanceof NotFoundError) {
+			return new Response('Not found', { status: 404 })
+		} else if (e instanceof MethodNotAllowedError) {
+			return new Response('Method not allowed', { status: 405 })
+		} else {
+			return new Response('Internal Error', { status: 500 })
+		}

-		throw err;
	}
}
```

The code in this PR is somewhat simpler, but is that behaviour documented somewhere?

vicentematus commented:
My bad, i was confused. Rich is right. I added the errors that are documented on the documentation of kv-asset-handler and fixed the changeset. Fixed.

benmccann commented on packages/adapter-cloudflare-workers/files/entry.js:
```diff
@@ -86,19 +91,29 @@ export default {
+		} else if (e instanceof MethodNotAllowedError) {
+			return new Response('Method not allowed', { status: 405 });
+		} else {
+			return new Response('Internal Error', { status: 500 });
```
it feels like this might make it hard for people to figure out what the underlying error is

vicentematus replied:
I cant see how to figure this out. Should it throw the error that it catches on the try catch clause🤔 ?

eltigerchino commented:
closing since the issue has been resolved. Completed by https://github.com/sveltejs/kit/pull/4276
